mlx/fast.cpp

```diff
@@ -127,4 +127,168 @@ bool RoPE::is_equivalent(const Primitive& other) const {
+    std::ostringstream msg;
+    msg << "[scaled_dot_product_attention] n_heads must be a multiple of n_kv_heads, found n_heads "
+        << n_q_heads << " for n_kv_heads " << n_kv_heads << ".";
+    throw std::invalid_argument(msg.str());
+  }
+
+  // q_seq_len == 1 for now in decoding (no mask) implementation
+  if (queries.shape(-2) > 1 && !mask.has_value()) {
+    std::ostringstream msg;
+    msg << "[scaled_dot_product_attention] query sequence length must be 1 at decode time, found  "
+        << queries.shape(-2) << ".";
+    throw std::invalid_argument(msg.str());
+  }
+
+  auto final_type = result_type({queries, keys, values});
+
+  auto q = astype(queries, final_type, s);
+  auto k = astype(keys, final_type, s);
+  auto v = astype(values, final_type, s);
+
+  auto out_shape =
+      std::vector<int>({q.shape(0), q.shape(1), q.shape(2), v.shape(-1)});
+
+  /* generic implementation for use cases that Metal implementation does not
+   * support. For non-supported cases listed below, use MLX primitives:
+   * * CPU implementation
+   * * batch size > 1
+   * * query sequence length > 1
+   * * non-null mask
+   */
+  bool needs_mask = mask.has_value();
+  auto fallback =
+      [scale, needs_mask, final_type, &s](const std::vector<array>& inputs) {
+        auto& q_tensor = inputs[0];
+        auto& k_tensor = inputs[1];
+        auto& v_tensor = inputs[2];
+        auto q_scaled = multiply(array(scale, q_tensor.dtype()), q_tensor, s);
+        int n_q_heads = q_tensor.shape()[1];
+        int n_kv_heads = k_tensor.shape()[1];
+
+        auto tile_if_needs_repeat = [](const array& arr,
+                                       int query_heads,
+                                       int kv_heads,
+                                       StreamOrDevice& s) -> array {
+          if (query_heads == kv_heads)
+            return arr;
+          int n_repeats = query_heads / kv_heads;
+          constexpr const int heads_axis =
+              1; // heads axis, assumes tensors arranged  as [0, 1, 2, 3] ->
+                 // [Batch, Heads, Sequence, Hidden]
+          auto ret = repeat(arr, n_repeats, heads_axis, s);
+          return ret;
+        };
+        auto k_tensor_tiled =
+            tile_if_needs_repeat(k_tensor, n_q_heads, n_kv_heads, s);
+        auto v_tensor_tiled =
+            tile_if_needs_repeat(v_tensor, n_q_heads, n_kv_heads, s);
+
+        // dim check on k, v; repeat if untiled, since naive matmul will have
+        // dim mismatch for GQA (MQA could make use of broadcast)
+        auto k_transposed = transpose(k_tensor_tiled, {0, 1, 3, 2}, s);
+        auto s_tensor = matmul(q_scaled, k_transposed, s);
+        if (needs_mask) {
+          auto mask_tensor = inputs[3];
+          s_tensor = add(s_tensor, mask_tensor, s);
+        }
+        auto p = softmax(s_tensor, std::vector<int>({-1}), s);
+        auto out_tensor = matmul(p, v_tensor_tiled, s);
+        out_tensor = astype(out_tensor, final_type, s);
```

Comment:
Interesting that you do the cast after the out projection here, was that intentional? In almost all cases we do it before the output projection. I think we should do that here as well.

Reply:
@bpkeene looking at this again, I think the right thing to do is the `softmax` in higher presicion and then downcast that to the final type. See how we do it e.g. [here](https://github.com/ml-explore/mlx-examples/blob/main/llms/mlx_lm/models/llama.py#L100). In almost all cases (training and inference) you want the softmax to be done in fp32. (We could always expose it as an option in the future, but for now I think having that as the default is good).

Reply:
sgtm! I'll add that to this change as well, along with downcasting afterwards for the output projection (in the event that we're using fp16)